### DISCUSSION TITLE
Read an existing DXF and change the font using .ttf file.

### DISCUSSION THREAD

**Discussion by joejoseph007** [mozman/ezdxf#918] [Q&A]
As the title says, I'm simply trying to read an existing dxf and change the font using a custom font and finally render the dxf using matplotlib (to either svg or png). Here's the code I'm using to render that, I'd appretiate any suggestions to solve this problem.

```Python
import matplotlib.pyplot as plt
import ezdxf

dxffile= # input dxf file
outfile= # output file (svg or png)

doc = ezdxf.readfile(dxffile)
msp = doc.modelspace()
auditor = doc.audit()
if len(auditor.errors) != 0:
    raise Exception(
        "This DXF document is damaged and can't be converted!")
else:
    ezdxf.addons.drawing.properties.MODEL_SPACE_BG_COLOR = '#FFFFFF'
    fig = plt.figure(dpi=600)
    ax = fig.add_axes([0, 0, 1, 1])
    ctx = RenderContext(doc)
    ctx.set_current_layout(msp)
    out = MatplotlibBackend(ax)
    Frontend(ctx, out).draw_layout(msp, finalize=True)
    fig.savefig(outfile)
```

**Comment by mozman** [comment]
First, don't change a global constant. Although Python does not support constants, uppercase variables should be treated as such, I guarantee this code will break in the future.

```Python
# don't do this
ezdxf.addons.drawing.properties.MODEL_SPACE_BG_COLOR = '#FFFFFF'
```

I'm sure you've seen this example on Stackoverflow and my comment on it is clear: https://stackoverflow.com/questions/63514133/is-there-a-way-to-change-the-background-color-in-ezdxf

This section of the documenation  shows how to do it right: https://ezdxf.mozman.at/docs/howto/drawing-addon.html#how-to-set-background-and-foreground-colors

```Python
from ezdxf.addons.drawing.properties import LayoutProperties

# -x-x-x snip -x-x-x-

msp_properties = LayoutProperties.from_layout(msp)
msp_properties.set_colors("#FFFFFF")
out = MatplotlibBackend(ax)

# override the layout properties and render the modelspace
Frontend(ctx, out).draw_layout(
    msp,
    finalize=True,
    layout_properties=msp_properties,
)
fig.savefig(outfile)
```

The first solution to this problem that comes to my mind is to change the font of all text styles:

```Python
for style in doc.styles:
    style.dxf.font = "Arial.ttf"
```

This does not work for font changes by inline code in MTEXT entities.

**Reply by joejoseph007** [reply]
Thank you for the sugestions!

Unfortunately, the solution for changing text does not work. 

I have attached a sample dxf whose font I'm trying to modify. The primary reason for changing the font is since most standard fonts don't have all the adequate Glyphs to render all the text.  

The DXF extension was modified since github doesn't support .dxf file attachments.

DXF.txt (DXF file removed)

**Comment by mozman** [comment]
Upload DXF files as zip-files and of course, you must use an appropriate font, in this case the unicode font "arialuni.ttf" does the job.

For Chinese texts, the font "simsun.ttc" can be used, the ".ttc" extensions is important, at least on my system.

**Reply by joejoseph007** [reply]
Thank you so much! This was a good solution.

I have a request though, I'm removing the dxf file, if you could also remove the screenshot it would be appreciated. Thanks!
